manim/mobject/geometry.py

```diff
@@ -863,6 +863,119 @@ def __init__(self, n=6, **kwargs):
+    **kwargs : Any
+        Affects how the ArcPolygon itself is drawn, but doesn't affect
+        passed arcs.
+
+    Attributes
+    ----------
+    arcs : :attr:`list`
+        The arcs used to initialize the ArcPolygon.
+
+    Examples
+    --------
+    One example of an arcpolygon is the Reuleaux triangle.
+    Instead of 3 straight lines connecting the outer points,
+    a Reuleaux triangle has 3 arcs connecting those points,
+    making a shape with constant width.
+
+    Passed arcs are stored as submobjects in the ArcPolygon.
+    This means that the arcs are changed along with the ArcPolygon,
+    for example when it's shifted, and these arcs can be manipulated
+    after the ArcPolygon has been initialized.
+
+    Also both the arcs contained in an ArcPolygon, as well as the
+    ArcPolygon itself are drawn, which affects draw time in ShowCreation
+    for example. In most cases the arcs themselves don't
+    need to be drawn, in which case they can be passed as invisible.
+        .. manim:: ArcPolygonExample
+
+            class ArcPolygonExample(Scene):
+                def construct(self):
+                    arc_conf = {"stroke_width":0}
+                    poly_conf = {"stroke_width":10,"stroke_color":BLUE,
+                      "fill_opacity":1,"color": PURPLE}
+                    a=[-1,0,0]
+                    b=[1,0,0]
+                    c=[0,np.sqrt(3),0]
+                    arc0=ArcBetweenPoints(a,b,radius=2,**arc_conf)
+                    arc1=ArcBetweenPoints(b,c,radius=2,**arc_conf)
+                    arc2=ArcBetweenPoints(c,a,radius=2,**arc_conf)
+                    reuleaux_tri=ArcPolygon(arc0,arc1,arc2,**poly_conf)
+                    self.play(FadeIn(reuleaux_tri))
+                    self.wait(2)
+
+    The ArcPolygon itself can also be hidden so that instead only the
+    contained arcs are drawn.
+        .. manim:: ArcPolygonExample2
+
+            class ArcPolygonExample2(Scene):
+                def construct(self):
+                    arc_conf = {"stroke_width":3,"stroke_color":RED,
+                      "fill_opacity":0.5,"color": GREEN}
+                    poly_conf = {"color":None}
+                    a=[-1,0,0]
+                    b=[1,0,0]
+                    c=[0,np.sqrt(3),0]
+                    arc0=ArcBetweenPoints(a,b,radius=2,**arc_conf)
+                    arc1=ArcBetweenPoints(b,c,radius=2,**arc_conf)
+                    arc2=ArcBetweenPoints(c,a,radius=2,**arc_conf)
+                    reuleaux_tri=ArcPolygon(arc0,arc1,arc2,**poly_conf)
+                    self.play(FadeIn(reuleaux_tri))
+                    self.wait(2)
+
+    ArcPolyon.arcs::
+        >>> from manim import ArcPolygon, Arc, ArcBetweenPoints
+        >>> ap=ArcPolygon(Arc(),ArcBetweenPoints([1,0,0],[0,1,0]),Arc())
+        >>> ap.arcs
+        [Arc, ArcBetweenPoints, Arc]
+    """
+
+    def __init__(self, *arcs, **kwargs):
+        if not all(isinstance(m, (Arc, ArcBetweenPoints)) for m in arcs):
```

Comment:
From a user point of view, I think it is rather unintuitive that `Polygon`s can be created by passing vertices, and `ArcPolygon`s are created by passing arcs. I think it would be more user-friendly if `ArcPolygon` could also be constructed by specifying its vertices, and then `__init__` could construct the corresponding `ArcBetweenPoints` or so.

Reply:
But how would that even work? You would have to do something weird like passing it alternating vertices and angles (which I consider even more confusing), or lists of vertices and angles (in which case you can pass it arcs directly again). Keep in mind that with vertices alone, it's not an ArcPolygon, it's just any old regular Polygon. The angles have to be user specified somehow. You'd have to give me some good idea for how to make that more intuitive than passing it arcs, I can't conceive of one.

Reply:
What speaks against a keyword `angle` which either constructs arcs having the same arc angle, or, as you say, a list of angles that needs to have the same number of entries as there are arcs? I would find that much more intuitive -- but that's my opinion; maybe others devs see this differently.

Reply:
Well, for one passing one single `angle` I reckon will be more of a niche case for using ArcPolygons overall. Sure, the simpler Reuleaux triangles (or pentagons etc) are a case of this, they have their use and are nice for a demonstration, but the vast majority of shapes will need different angles. With the exception of these "fat" and "slim" triangles every other time I used this class was with multiple varying angles. Every request needing this kind of class on Discord I saw needed varying angles.

Second, `ArcBetweenPoints` now has a `radius` trait too, so you'd have to be able to use that as well.
This alone means you could instantiate an `ArcPolygon` in 3 general ways, passing arcs, passing points+angle and passing points+radius. If you pass it nested lists then you basically just replace an arc you pass with something like `[[x,y,z],[x,y,z],angle=p]` and maybe then `[[x,y,z],[x,y,z],radius=q]`. That's already kinda messy.

Third, there's also configs per arcs, which you can do by simply configuring the arcs you pass currently, but if you'd also add this ability to passed nested lists that'd be... messy. Like really messy and complicated. Frankly as a user, especially a new one that'd intimidate me more than it'd help.

Fourth, from experience I can say that setting up proper arcpolygons just simply tends to be a finicky process, and I've learned a lot about how to work with them more efficiently and cleanly, which I have poured into the examples. I did the whole "lists flying around everywhere"-thing in my 3rd video, and also dirty quick instantiation methods, and I would strongly advise against it in hindsight. Adding a meta-layer to arc creation by having an `ArcPolygon` internal nested list format means users likely will have to fight a battle on two fronts anyway, `Arc` and `ArcBetweenPoints` on one side, and then the custom `ArcPolygon` format on the other side.
Having a solid understanding and direct control and feedback of how to create which arcs goes a mile and a half in making proper arcpolygons without tearing out your hair.

The work process I envision for creating an `ArcPolygon`, especially complex shapes is this:
1. Make a list of the points you want to have, and ideally some design dicts. So far, so easy.
2. Make the arcs. This can already be complicated enough because figuring out the right angles (or radii) is more a less a challenge depending on your desired shape. 
3.  Setup the ArcPolygon with just the arcs visible first. If everything is fine, switch to making the arcs invisible and the `ArcPolygon` visible and be done with it. If not, you have your points and arc definitions easily accessible for quick tinkering, and you can give the arcs individual appearances to see precisely where the problem is.

Anyway, I hope this WoT explains my design decision. I realized I should adjust the example on one small front a little though.

Reply:
Thank you for your thorough explanation, I can understand where you are coming from now a bit better, and can understand your design choices.

I still think that the discrepancy between `Polygon` and `ArcPolygon` is unfortunate; maybe I am also just not a big fan of an interface requiring me to setup (complex) objects (correctly) before I am able to use it.

That being said, I do think that an interface that allows for each of the following scenarios would be nice:

```
>>> ArcPolygon(a, b, c, angle=15*DEGREE)
>>> ArcPolygon(a, b, c, radius=0.42)
>>> ArcPolygon(a, b, c, arc_configs=[{'radius': 0.7, 'color': RED}, {'angle': PI/8}, {'segment_height': 1, 'someotherarcproperty': 42}])
```
where a,b,c are points.

(Being able to specify the segment height when constructing an `ArcBetweenPoints` seems generally useful to me -- you are experienced with these, what do you think? Also, sorry if this is already possibly; I am on mobile and checking was too difficult.)

But in any case, this is just my opinion; my concerns regarding the discrepancy with `Polygon`s interface could also be resolved by renaming `ArcPolygon` to, like, `PolygonFromArcs` or so. It would be interesting to hear what other devs think about this.

Reply:
I've slept on this for a night.
> But in any case, this is just my opinion; my concerns regarding the discrepancy with `Polygon`s interface could also be resolved by renaming `ArcPolygon` to, like, `PolygonFromArcs` or so. It would be interesting to hear what other devs think about this.

While I get where you're coming from, I'm very strongly against this renaming, here's why: These shapes are actually called arcpolygons, and for a a good reason, since that's what they are. They are more general than your run of the mill polygon and less general than just a closed curve, placing them in between. They have a lot of arcpolygon specific, interesting and useful math to themselves. However if you'll search across the internet you'll have a *REALLY* hard time finding any proper documentation on this. The wikipedia page for Reuleaux triangles only references the much more general concept of a closed curve and the much more specific Reuleaux polygons, leaving a pretty wide gap inbetween, which is pretty bad. Strangely enough there is a german wiki page for arcpolygons (https://de.wikipedia.org/wiki/Bogenvieleck), but even that page is honestly kind of wrong, and in english you'll just about find them used in a few select research papers.
When I was first looking into these shapes for the Hadwiger-Nelson problem it was so insanely hard to find stuff on arcpolygons I ended up reinventing the entire wheel from scratch. That's a problematic precedent I would really like to avoid factoring into. Especially because this class of shapes is actually all around us constantly.

So while there is some discrepany with `Polygon`s interface, I think accurate usage of mathematical terms is more relevant. Especially if it's one of the most underused and underexplained terms, it's better to not pile some more abuse onto it. An `ArcPolygon` is a universally more generalized version of a `Polygon`, so it's just to be expected there will be discrepancys with the interface.

> That being said, I do think that an interface that allows for each of the following scenarios would be nice:

Nice for sure, but still a niche case. What I think would be best here is to have `ArcPolygon` the way it is, staying true to the mathematical terminology, and maybe a different class allowing for these niche specific cases. Problem is I have no idea what it'd be called.
Sure, I could also put this all into `ArcPolygon`, but both the resulting code as well as the documentation would probably be one of the most intimidating things across our entire codebase. I don't know if that's really how we want to play it. Splitting it into 2 classes would help at least a bit. Any naming idea? Just plain old `ArcPolygon2` seems off-base for obvious reason. `ArcPolygonFromPoints` maybe...

> Being able to specify the segment height when constructing an ArcBetweenPoints seems generally useful to me -- you are experienced with these, what do you think?

Yeah that might be useful, it's another valid way to make arcs. I added radius as an option in the first place because it makes defining certain ArcPolygons I need way easier. I'll make a note to possibly add this in the future.

On a side note I realized we might want to add a `ClosedCurve` class later on as well. And as a yet more general case, its interface would also inevitably end up being different from `ArcPolygon` in turn.

Reply:
> I've slept on this for a night.
> 
> > But in any case, this is just my opinion; my concerns regarding the discrepancy with `Polygon`s interface could also be resolved by renaming `ArcPolygon` to, like, `PolygonFromArcs` or so. It would be interesting to hear what other devs think about this.
> 
> While I get where you're coming from, I'm very strongly against this renaming,

You are right, `PolygonFromArcs` is not ideal, and semantically plain wrong. It was just a suggestion made very late at night. :-)

> However if you'll search across the internet you'll have a _REALLY_ hard time finding any proper documentation on this. The wikipedia page for Reuleaux triangles only references the much more general concept of a closed curve and the much more specific Reuleaux polygons, leaving a pretty wide gap inbetween, which is pretty bad. Strangely enough there is a german wiki page for arcpolygons (https://de.wikipedia.org/wiki/Bogenvieleck), but even that page is honestly kind of wrong, and in english you'll just about find them used in a few select research papers.

If there is a reference properly definining arc polygons, then we can and should include a link to it.

> So while there is some discrepany with `Polygon`s interface, I think accurate usage of mathematical terms is more relevant. Especially if it's one of the most underused and underexplained terms, it's better to not pile some more abuse onto it. An `ArcPolygon` is a universally more generalized version of a `Polygon`, so it's just to be expected there will be discrepancys with the interface.

If I understand correctly, an arc polygon is a closed curve consisting of arc segments. I don't see how the interface
```py
ArcPolygon(*vertices, *arc_configs)
```
misuses mathematical terms, and
```py
ArcPolygon(*arcs)
```
doesn't.

> > That being said, I do think that an interface that allows for each of the following scenarios would be nice:
> 
> Nice for sure, but still a niche case.

I don't see what is not covered. If we allow to pass a list of dicts specifying construction parameters for the corresponding `ArcBetweenPoints` as I tried to illustrate with my last example, you have the same functionality as if you would construct the `Arc`s yourself before -- or am I missing something?

> Sure, I could also put this all into `ArcPolygon`, but both the resulting code as well as the documentation would probably be one of the most intimidating things across our entire codebase. 

Yes, I agree. We definitely shouldn't do that.

> I don't know if that's really how we want to play it. Splitting it into 2 classes would help at least a bit. Any naming idea? Just plain old `ArcPolygon2` seems off-base for obvious reason. `ArcPolygonFromPoints` maybe...

I actually really like the idea of splitting it into two separate classes: what about `ArcPolygon` (for something that behaves similar to `Polygon`, and `ArcPolygonFromArcs` for the current `ArcPolygon` implementation? If this is an acceptable compromise for you and if you would like me to, I'd be happy to help implementing `ArcPolygon`.

> On a side note I realized we might want to add a `ClosedCurve` class later on as well. And as a yet more general case, its interface would also inevitably end up being different from `ArcPolygon` in turn.

Of course, but a closed curve can in general no longer be generalized as a curve whose segments are something simple.

Reply:
With mathematical terms I was only referring to `PolygonFromArcs` specifically. Your suggested way of creating arcpolygons is perfectly sound, mathematically speaking.
> If there is a reference properly definining arc polygons, then we can and should include a link to it.

I wish. As far as I can tell after days of searching, there isn't, so... yeah.

> I don't see what is not covered. If we allow to pass a list of dicts specifying construction parameters for the corresponding ArcBetweenPoints as I tried to illustrate with my last example, you have the same functionality as if you would construct the Arcs yourself before -- or am I missing something?

Nope, this was just in reference to your idea of passing points and one trait. Probably should've clarified that.

> I actually really like the idea of splitting it into two separate classes: what about ArcPolygon (for something that behaves similar to Polygon, and ArcPolygonFromArcs for the current ArcPolygon implementation? If this is an acceptable compromise for you and if you would like me to, I'd be happy to help implementing ArcPolygon.

While I'm mildly against making "from points" to be the standard variant because it encourages users to learn additional new arcpolygon specific syntax and discourages them from working with the existing arcs they will with decent probability need somewhere close by anyway if they're alread arcpolygons, it's a compromise I still can get behind I guess. If so then yes, it'd be good if you could take up the flag on the from points variant. Because they work very differently either class likely won't inherit from the other one either.
> Of course, but a closed curve can in general no longer be generalized as a curve whose segments are something simple.

For sure, but I have ideas how to implement it. Not uh... this week though. Or next week. I'll make an issue and tag myself.

Reply:
I have added a suggestion for an example via 0ab5e39f, I've named the class `ArcPolygonP` in order not to interfere with your `ArcPolygon` class. I'd still be in favor or renaming, though.